src/Servers/Kestrel/Core/src/Internal/Http3/QPack/QPackEncoder.cs

```diff
@@ -0,0 +1,504 @@
+                    {
+                        throw new QPackEncodingException("TODO sync with corefx" /* CoreStrings.HPackErrorNotEnoughBuffer */);
+                    }
+                    return false;
+                }
+
+                length += headerLength;
+            } while (_enumerator.MoveNext());
+
+            return true;
+        }
+
+        private bool EncodeHeader(string name, string value, Span<byte> buffer, out int length)
+        {
+            var i = 0;
+            length = 0;
+
+            if (buffer.Length == 0)
+            {
+                return false;
+            }
+
+            if (!EncodeNameString(name, buffer.Slice(i), out var nameLength, lowercase: true))
+            {
+                return false;
+            }
+
+            i += nameLength;
+
+            if (i >= buffer.Length)
+            {
+                return false;
+            }
+
+            if (!EncodeValueString(value, buffer.Slice(i), out var valueLength, lowercase: false))
+            {
+                return false;
+            }
+
+            i += valueLength;
+
+            length = i;
+            return true;
+        }
+
+        private bool EncodeValueString(string s, Span<byte> buffer, out int length, bool lowercase)
+        {
+            const int toLowerMask = 0x20;
+
+            var i = 0;
+            length = 0;
+
+            if (buffer.Length == 0)
+            {
+                return false;
+            }
+
+            buffer[0] = 0;
+
+            if (!IntegerEncoder.Encode(s.Length, 7, buffer, out var nameLength))
+            {
+                return false;
+            }
+
+            i += nameLength;
+
+            // TODO: use huffman encoding
+            for (var j = 0; j < s.Length; j++)
+            {
+                if (i >= buffer.Length)
+                {
+                    return false;
+                }
+
+                buffer[i++] = (byte)(s[j] | (lowercase && s[j] >= (byte)'A' && s[j] <= (byte)'Z' ? toLowerMask : 0));
+            }
+
+            length = i;
+            return true;
+        }
+
+        private bool EncodeNameString(string s, Span<byte> buffer, out int length, bool lowercase)
+        {
+            const int toLowerMask = 0x20;
+
+            var i = 0;
+            length = 0;
+
+            if (buffer.Length == 0)
+            {
+                return false;
+            }
+
+            buffer[0] = 0x30;
+
+            if (!IntegerEncoder.Encode(s.Length, 3, buffer, out var nameLength))
+            {
+                return false;
+            }
+
+            i += nameLength;
+
+            // TODO: use huffman encoding
+            for (var j = 0; j < s.Length; j++)
+            {
+                if (i >= buffer.Length)
+                {
+                    return false;
+                }
+
+                buffer[i++] = (byte)(s[j] | (lowercase && s[j] >= (byte)'A' && s[j] <= (byte)'Z' ? toLowerMask : 0));
```

Comment:
_Name_ is a header name? If so this can be simplified (see above):
```suggestion
                buffer[i++] = (byte)(s[j] | (lowercase ? toLowerMask : 0));
```